packages/fulfillment/src/services/fulfillment-module-service.ts

```diff
@@ -151,7 +808,130 @@ export default class FulfillmentModuleService<
   ): Promise<
     FulfillmentTypes.ServiceZoneDTO[] | FulfillmentTypes.ServiceZoneDTO
   > {
-    return []
+    const updatedServiceZones = await this.updateServiceZones_(
+      data,
+      sharedContext
+    )
+
+    return await this.baseRepository_.serialize<
+      FulfillmentTypes.ServiceZoneDTO | FulfillmentTypes.ServiceZoneDTO[]
+    >(updatedServiceZones, {
+      populate: true,
+    })
+  }
+
+  protected async updateServiceZones_(
+    data:
+      | FulfillmentTypes.UpdateServiceZoneDTO[]
+      | FulfillmentTypes.UpdateServiceZoneDTO,
+    sharedContext?: Context
+  ): Promise<TServiceZoneEntity | TServiceZoneEntity[]> {
+    const data_ = Array.isArray(data) ? data : [data]
+
+    const geoZonesToCreate: FulfillmentTypes.CreateGeoZoneDTO[] = []
+    const serviceZoneGeoZonesMap = new Map<
+      string,
+      Map<string, FulfillmentTypes.CreateGeoZoneDTO | { id: string }>
+    >()
+
+    const existingGeoZones: TGeoZoneEntity[] = []
+    let existingGeoZonesMap = new Map<string, TGeoZoneEntity>()
+
+    const geoZoneIds: string[] = []
+
+    data_.forEach((serviceZone) => {
+      if ("geo_zones" in serviceZone && serviceZone.geo_zones) {
+        const geo_zones = serviceZone.geo_zones
+
+        geo_zones.forEach((geoZone) => {
+          if ("id" in geoZone) {
+            geoZoneIds.push(geoZone.id)
+          }
+        })
+      }
+    })
+
+    if (geoZoneIds.length) {
+      existingGeoZones.push(
+        ...(await this.geoZoneService_.list(
+          {
+            id: geoZoneIds,
+          },
+          {},
+          sharedContext
+        ))
+      )
+      existingGeoZonesMap = new Map(
+        existingGeoZones.map((geoZone) => [geoZone.id, geoZone])
+      )
+    }
+
+    data_.forEach((serviceZone) => {
+      if ("geo_zones" in serviceZone && serviceZone.geo_zones) {
+        const geo_zones = serviceZone.geo_zones
+
+        const geoZoneTuple: [
+          string,
+          FulfillmentTypes.CreateGeoZoneDTO | { id: string }
+        ][] = geo_zones.map((geoZone) => {
+          if (!("id" in geoZone)) {
+            geoZonesToCreate.push(geoZone)
+          }
+
+          const existingZone =
+            "id" in geoZone ? existingGeoZonesMap.get(geoZone.id)! : geoZone
```

Comment:
question: Even though we queried the database for GeoZones, couldn't we be in a situation where the ID didn't exist in the database i.e. the map would not contain the GeoZone? In that case, `existingZone` will be undefined. 

This comment is relevant for all places where we have this pattern of "query the DB and add them to a map for later use"

Reply:
You are right, i can have an extra step that query then do a diff to ensure that 💪 ill add that in